nemo/collections/nlp/models/language_modeling/megatron_glue_model.py

```diff
@@ -0,0 +1,269 @@
+
+from nemo.collections.common.metrics.classification_accuracy import ExactStringMatchMetric
+from nemo.collections.nlp.data.glue_benchmark.glue_benchmark_dataset import TextToTextGLUEDataset
+from nemo.collections.nlp.models.language_modeling.megatron_t5_model import MegatronT5Model
+from nemo.collections.nlp.models.nlp_model import NLPModel
+from nemo.collections.nlp.modules.common.megatron.utils import average_losses_across_data_parallel_group
+from nemo.utils import logging
+
+__all__ = ['MegatronT5FineTuneModel']
+
+
+class MegatronT5FineTuneModel(NLPModel):
+    """
+    Megatron T5 finetuning
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        super().__init__(cfg, trainer)
+        # TODO: Fix this once apex patches FusedScaledMaskedSoftmax.
+        # This is a workaround for the fact that `masked_softmax_fusion` has issues with certain input sizes that may be present while finetuning.
+        t5_cfg = MegatronT5Model.restore_from(
+            self.register_artifact('t5_base_model', cfg.restore_from_path), trainer=trainer, return_config=True
+        )
+        OmegaConf.set_struct(t5_cfg, True)
+        with open_dict(t5_cfg):
+            t5_cfg.masked_softmax_fusion = False
+
+        self.model = MegatronT5Model.restore_from(
+            self.register_artifact('t5_base_model', cfg.restore_from_path),
+            trainer=trainer,
+            override_config_path=t5_cfg,
+        )
+
+    def training_step(self, batch, batch_idx):
+        pass
+
+    def validation_step(self, batch, batch_idx):
+        pass
+
+    def validation_epoch_end(self, outputs):
+        pass
+
+    def process_batch(self, batch):
+        return self.model.process_batch(batch)
+
+    def build_train_valid_test_datasets(self):
+        pass
+
+    def build_pretraining_data_loader(self, dataset):
+        pass
+
+    def setup(self, stage=None):
+        pass
+
+    def setup_training_data(self):
+        pass
+
+    def setup_validation_data(self):
+        pass
+
+    def setup_test_data(self):
+        pass
+
+    def on_before_optimizer_step(self, optimizer, optimizer_idx):
+        """PTL hook that is called after unscaling gradients when using native amp.
+           We use gradient clipping implementation from megatron-lm.
+        """
+        pass
+
+    def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int] = None) -> Any:
+        return self.model.prediction_step(batch, batch_idx, dataloader_idx)
+
+
+class MegatronT5GLUEModel(MegatronT5FineTuneModel):
+    """
+    Megatron T5 finetuning for GLUE
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        super().__init__(cfg=cfg, trainer=trainer)
+        self.cfg = cfg
+        self.acc_metric = ExactStringMatchMetric()
+
+    def training_step(self, batch, batch_idx):
+        tokens_enc, tokens_dec, loss_mask, labels, enc_mask, dec_mask, enc_dec_mask = self.process_batch(batch)
```

Comment:
Is it really needed? Each process should get a copy of data from its dataloader. No need to `broadcast` right?

Reply:
Yeah, right now I'm just using process_batch to move things to torch tensors.